rclpy/src/rclpy/subscription.cpp

```diff
@@ -0,0 +1,162 @@
+    PyErr_WarnFormat(
+      PyExc_RuntimeWarning, stack_level, "_rclpy_destroy_subscription got NULL pointer");
+    return;
+  }
+
+  rcl_ret_t ret = rcl_subscription_fini(&(sub->subscription), sub->node);
+  if (RCL_RET_OK != ret) {
+    // Warning should use line number of the current stack frame
+    int stack_level = 1;
+    PyErr_WarnFormat(
+      PyExc_RuntimeWarning, stack_level, "Failed to fini subscription: %s",
+      rcl_get_error_string().str);
+  }
+  PyMem_Free(sub);
+}
+
+py::capsule
+subscription_create(
+  py::capsule pynode, py::object pymsg_type, std::string topic,
+  py::capsule pyqos_profile)
+{
+  auto node = static_cast<rcl_node_t *>(
+    rclpy_handle_get_pointer_from_capsule(pynode.ptr(), "rcl_node_t"));
+  if (!node) {
+    throw py::error_already_set();
+  }
+
+  auto msg_type = static_cast<rosidl_message_type_support_t *>(
+    rclpy_common_get_type_support(pymsg_type.ptr()));
+  if (!msg_type) {
+    throw py::error_already_set();
+  }
+
+  rcl_subscription_options_t subscription_ops = rcl_subscription_get_default_options();
+
+  if (!pyqos_profile.is(py::none())) {
+    if (0 != strcmp("rmw_qos_profile_t", pyqos_profile.name())) {
+      throw py::value_error("capsule is not an rmw_qos_profile_t");
+    }
+    auto qos_profile = static_cast<rmw_qos_profile_t *>(pyqos_profile);
+    subscription_ops.qos = *qos_profile;
+  }
+
+  // Use smart pointer to make sure memory is free'd on error
+  auto deleter = [](rclpy_subscription_t * ptr) {_rclpy_destroy_subscription(ptr);};
+  auto sub = std::unique_ptr<rclpy_subscription_t, decltype(deleter)>(
+    static_cast<rclpy_subscription_t *>(PyMem_Malloc(sizeof(rclpy_subscription_t))),
+    deleter);
+  if (!sub) {
+    throw std::bad_alloc();
+  }
+  sub->subscription = rcl_get_zero_initialized_subscription();
+  sub->node = node;
+
+  rcl_ret_t ret = rcl_subscription_init(
+    &(sub->subscription), node, msg_type,
+    topic.c_str(), &subscription_ops);
+  if (ret != RCL_RET_OK) {
+    if (ret == RCL_RET_TOPIC_NAME_INVALID) {
+      std::string error_text{"Failed to create subscription due to invalid topic name '"};
+      error_text += topic;
+      error_text += "'";
+      throw RCLError(error_text);
+    }
+    throw RCLError("Failed to create subscription");
+  }
+
+  PyObject * pysub_c =
+    rclpy_create_handle_capsule(sub.get(), "rclpy_subscription_t", _rclpy_destroy_subscription);
+  if (!pysub_c) {
+    throw py::error_already_set();
+  }
+  auto pysub = py::reinterpret_steal<py::capsule>(pysub_c);
+  // pysub now owns the rclpy_subscription_t
+  sub.release();
+
+  auto sub_handle = static_cast<rclpy_handle_t *>(pysub);
+  auto node_handle = static_cast<rclpy_handle_t *>(pynode);
+  _rclpy_handle_add_dependency(sub_handle, node_handle);
+  if (PyErr_Occurred()) {
+    _rclpy_handle_dec_ref(node_handle);
+    throw py::error_already_set();
+  }
+
+  return pysub;
+}
+
+py::object
+subscription_get_logger_name(py::capsule pysubscription)
```

Comment:
Returning a `py::object` was the only way I could figure out to return `None` from a function that normally returns a string. Is there a better way to do this?

Reply:
Maybe `std::optional` from C++17? Pybind11 claims it supports it: https://pybind11.readthedocs.io/en/stable/advanced/cast/stl.html#c-17-library-containers . Otherwise, the way it's done here LGTM

Reply:
This package is still targeting C++14. Maybe we can consider using std::optional if/when we bump that.